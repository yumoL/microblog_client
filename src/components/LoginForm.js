/**
 * @description login form
 */
import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Grid, Header, Form, Segment, Button, Message } from 'semantic-ui-react'
import { Link, withRouter, Redirect } from 'react-router-dom'
import { useField } from '../hooks/useField'
import { login } from '../reducers/userReducer'
import { setNotification } from '../reducers/notificationReducer'

const LoginForm = props => {
  const userName = useField('text')
  const password = useField('password')

  const loginButtonDisabled = userName.value.length === 0 || password.value.length === 0

  const handleSubmit = async e => {
    e.preventDefault()
    const user = {
      userName: userName.value,
      password: password.value
    }
    try {
      await props.login(user)
    } catch (e) {
      props.setNotification({ message: 'Login failed, wrong username or password', error: true })
    }
  }

  const { from } = props.location.state || { from: { pathname: '/home' } }
  if (props.user.token) {
    props.setNotification({ message: 'You have logged in', error: false })
    return (
      <div className='redirect'>
        <Redirect to={from} />
      </div>

    )
  }

  return (
    <div>
      <Grid textAlign='center' style={{ height: '100vh' }} verticalAlign='middle'>
        <Grid.Column style={{ maxWidth: 450 }}>
          <Header as='h2' color='teal' textAlign='center'>
            Log in to your account
          </Header>
          <Form size='large' onSubmit={handleSubmit}>
            <Segment stacked>
              <Form.Input className='userName' fluid icon='user' iconPosition='left' placeholder='User Name' {...userName} required />
              <Form.Input
                className='pwd'
                fluid
                icon='lock'
                placeholder='Password'
                {...password}
                required
              />
              <Button type='submit' color='teal' fluid size='large' disabled={loginButtonDisabled}>
                Login
              </Button>
            </Segment>
          </Form>
          <Message>
            New to us? <Link to='/register'>Sign Up</Link>
          </Message>
        </Grid.Column>
      </Grid>

    </div>

  )
}

const mapStateToProps = state => {
  return {
    user: state.user
  }
}

const mapDispatchToProps = {
  login,
  setNotification
}

LoginForm.propTypes = {
  history: PropTypes.object.isRequired,
  login: PropTypes.func.isRequired,
  setNotification: PropTypes.func.isRequired,
  user: PropTypes.object,
  location: PropTypes.object.isRequired
}

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(LoginForm)
)