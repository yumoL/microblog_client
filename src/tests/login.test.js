/* eslint-disable no-undef */
import React from 'react'
import { fireEvent, wait } from '@testing-library/react'

jest.mock('../services/user')

import LoginForm from '../components/LoginForm'
import { renderWithReduxAndRouter } from './testHelper'

describe('Login', () => {

  let Wrapper

  beforeEach(() => {
    Wrapper = renderWithReduxAndRouter(<LoginForm />)
  })

  test('should render login form correctly', () => {
    expect(Wrapper).toMatchSnapshot()
  })

  test('Login button is disabled when nothing has been inputted', () => {
    const button = Wrapper.getByText('Login')
    expect(button).toBeDisabled()
  })

  test('Redirect to home page after login', async () => {
    const userNameInput = Wrapper.container.querySelector('#userName')
    const pwdInput = Wrapper.container.querySelector('#pwd')
    const form = Wrapper.container.querySelector('form')
    fireEvent.change(userNameInput, { target: { value: 'user1' } })
    fireEvent.change(pwdInput, { target: { value: '123' } })
    fireEvent.submit(form)

    // await wait(() => expect(Wrapper.container.querySelector('.redirect')).toBeInTheDocument())
    // expect(Wrapper.container.querySelector('#userName')).not.toBeTruthy()
  })

})