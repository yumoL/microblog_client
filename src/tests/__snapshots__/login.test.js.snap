// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render login form correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="ui center aligned middle aligned grid"
          style="height: 100vh;"
        >
          <div
            class="column"
            style="max-width: 450px;"
          >
            <h2
              class="ui teal center aligned header"
              id="loginHeader"
            >
              Log in to your account
            </h2>
            <form
              class="ui large form"
            >
              <div
                class="ui stacked segment"
              >
                <div
                  class="required field"
                >
                  <div
                    class="ui fluid left icon input"
                  >
                    <input
                      id="userName"
                      placeholder="User Name"
                      required=""
                      type="text"
                      value=""
                    />
                    <i
                      aria-hidden="true"
                      class="user icon"
                    />
                  </div>
                </div>
                <div
                  class="required field"
                >
                  <div
                    class="ui fluid icon input"
                  >
                    <input
                      id="pwd"
                      placeholder="Password"
                      required=""
                      type="password"
                      value=""
                    />
                    <i
                      aria-hidden="true"
                      class="lock icon"
                    />
                  </div>
                </div>
                <button
                  class="ui teal large fluid disabled button"
                  disabled=""
                  id="loginButton"
                  tabindex="-1"
                  type="submit"
                >
                  Login
                </button>
              </div>
            </form>
            <div
              class="ui message"
            >
              New to us? 
              <a
                href="/register"
              >
                Sign Up
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="ui center aligned middle aligned grid"
        style="height: 100vh;"
      >
        <div
          class="column"
          style="max-width: 450px;"
        >
          <h2
            class="ui teal center aligned header"
            id="loginHeader"
          >
            Log in to your account
          </h2>
          <form
            class="ui large form"
          >
            <div
              class="ui stacked segment"
            >
              <div
                class="required field"
              >
                <div
                  class="ui fluid left icon input"
                >
                  <input
                    id="userName"
                    placeholder="User Name"
                    required=""
                    type="text"
                    value=""
                  />
                  <i
                    aria-hidden="true"
                    class="user icon"
                  />
                </div>
              </div>
              <div
                class="required field"
              >
                <div
                  class="ui fluid icon input"
                >
                  <input
                    id="pwd"
                    placeholder="Password"
                    required=""
                    type="password"
                    value=""
                  />
                  <i
                    aria-hidden="true"
                    class="lock icon"
                  />
                </div>
              </div>
              <button
                class="ui teal large fluid disabled button"
                disabled=""
                id="loginButton"
                tabindex="-1"
                type="submit"
              >
                Login
              </button>
            </div>
          </form>
          <div
            class="ui message"
          >
            New to us? 
            <a
              href="/register"
            >
              Sign Up
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
